internal/pkg/infrastructure/postgres/reading.go

```diff
@@ -6,78 +6,407 @@
+		var reading model.Reading
+
+		readingDBModel, err := pgx.RowToStructByNameLax[dbModels.Reading](row)
+		if err != nil {
+			return nil, pgClient.WrapDBError("failed to convert row to map", err)
+		}
+
+		// convert the BaseReading fields to BaseReading struct defined in contract
+		baseReading := readingDBModel.GetBaseReading()
+
+		valueType := baseReading.ValueType
+
+		if valueType == common.ValueTypeBinary {
+			// reading type is BinaryReading
+			binaryReading := model.BinaryReading{
+				BaseReading: baseReading,
+				MediaType:   *readingDBModel.MediaType,
+				BinaryValue: readingDBModel.BinaryValue,
+			}
+			reading = binaryReading
+		} else if baseReading.ValueType == common.ValueTypeObject {
+			// reading type is ObjectReading
+			objReading := model.ObjectReading{
+				BaseReading: baseReading,
+				ObjectValue: readingDBModel.ObjectValue,
+			}
+			reading = objReading
+		} else {
+			// reading type is SimpleReading
+			simpleReading := model.SimpleReading{
+				BaseReading: baseReading,
+				Value:       *readingDBModel.Value,
+			}
+			reading = simpleReading
+		}
+
+		return reading, nil
+	})
+
+	if err != nil {
+		return nil, errors.NewCommonEdgeXWrapper(err)
+	}
+
+	return readings, nil
+}
+
+// deleteReadings delete the data rows with given sql statement and passed args
+func deleteReadings(ctx context.Context, tx pgx.Tx, args ...any) errors.EdgeX {
+	sqlStatement := sqlDeleteByColumn(readingTableName, eventIdFKCol)
+	commandTag, err := tx.Exec(
+		ctx,
+		sqlStatement,
+		args...,
+	)
+	if commandTag.RowsAffected() == 0 {
+		return errors.NewCommonEdgeX(errors.KindContractInvalid, "no reading found", nil)
+	}
+	if err != nil {
+		return pgClient.WrapDBError("reading(s) delete failed", err)
+	}
+	return nil
+}
+
+// deleteReadings delete the readings with event_id in the range of the sub query
+func deleteReadingsBySubQuery(ctx context.Context, tx pgx.Tx, subQuerySql string, args ...any) errors.EdgeX {
+	sqlStatement := sqlDeleteByColumn(readingTableName, eventIdFKCol)
+	subQueryCond := "ANY ( " + subQuerySql + " )"
+	sqlStatement = strings.Replace(sqlStatement, "$1", subQueryCond, 1)
+
+	commandTag, err := tx.Exec(
+		ctx,
+		sqlStatement,
+		args...,
+	)
+	if commandTag.RowsAffected() == 0 {
+		return errors.NewCommonEdgeX(errors.KindContractInvalid, "no reading found", nil)
+	}
+	if err != nil {
+		return pgClient.WrapDBError("reading(s) delete failed", err)
+	}
+	return nil
+}
+
+// addReadingsInTx converts reading interface to BinaryReading/ObjectReading/SimpleReading structs first based on the reading value type
+// and then perform the CopyFromSlice transaction to insert readings in batch
+func addReadingsInTx(tx pgx.Tx, readings []model.Reading, eventId string) error {
+	var readingDBModels []dbModels.Reading
+
+	for _, r := range readings {
+		baseReading := r.GetBaseReading()
+		var readingDBModel dbModels.Reading
+		valueType := baseReading.ValueType
+		if valueType == common.ValueTypeBinary {
```

Comment:
instead of checking valueType and do the conversion, you can just use type switch case without checking the valueType here
```
switch v := r.(type) {
	case model.BinaryReading:
...
```